### ISSUE TITLE
Sykube and Singularity-cri updates for Linux kernel 4+

### ISSUE THREAD

**Issue by Nienkeeijsvogel** [Nienkeeijsvogel/sykube#1]
Hi everyone,

Sykube and the container runtime interface for Kubernetes, Singularity-cri, were not working on newer Linux Kernels (Ubuntu 18.04 and newer , Centos8 and newer etc). I adjusted the source code of both, Sykube and Singularity-cri, and they are now operative on the newest operating systems as ubuntu 22.04 and centos9. Since both Linux kernels and Singularity versions keep on being updated, Sykube and Singularity-cri require maintenance. The switch to cgroup v2 is something that still needs to be worked on so Ubuntu 22.04 requires switching to the legacy cgroup. If there are any more questions or interest in my forks please comment below. 

Nienke

**Comment by vsoch** [comment]
Thanks @Nienkeeijsvogel ! Could you give a high level overview of these components in Kubernetes for the new developer? I've worked on Operators but I haven't worked on any CRI. I think a high level description and some links to docs or code would be super helpful.

**Comment by Nienkeeijsvogel** [comment]
@vsoch The branch v1.0.0-beta.8 of https://github.com/Nienkeeijsvogel/singularity-cri
is the operative branch atm.
Important for the singularity-cri is that while it requires having Singularity installed its not just a container runtime interface that translates Kubernetes requests and passes them to the Singularity executable;  Singularity-cri does part of the execution as well and relies on importing Singularity in the go.mod file for this. This version is equally (if not more) important for the implementation of the Singularity-cri. Since it imports functions from Singularity by these statements at the top of files: "github.com/sylabs/singularity/pkg/image"
"github.com/sylabs/singularity/pkg/signing"
What i think would be the most convenient way for updating the Singularity-cri, is to gradually update Kubernetes and Singularity versions with the 'go mod update package command' and adjust and restructure the source code accordingly. 
If you try to build, you will get errors which will guide you well in the process and will point to changes in underlying versions.
Gradually updating versions will make the process a bit more straightforward because you can look at change logs for versions for Kubernetes and Singularity. 
By working on Singularity-cri you get familiar with Kubernetes, it requests and the Singularity-cri source code pretty easily. 
If you have any more questions, i would be happy to elaborate more and go a bit more into detail on the source code.

**Comment by vsoch** [comment]
Gotcha! Is there enough working now so it will run with something like minikube? If you can point me to those docs and I can see how it works, that's mostly what I need to help working on it. 

Thanks for this detail!

**Comment by Nienkeeijsvogel** [comment]
@vsoch Yes Sykube works now the same way Minikube does! I updated the README.md file of my fork so it should be clear now on how to install on older/newer operating systems. Sykube itself is straightforward; the .def file is the definition file for the container in which the Kubernetes cluster resides. the Sykube file will be copied to /usr/local/bin and is a sh-scripted file that contains runtime functions for Sykube, i.e. starting,stopping a cluster. This file also changes certain settings as iptables, firewalls to ensure everything can run properly.

**Comment by Nienkeeijsvogel** [comment]
@vsoch Hi Vanessa, let me know if you run into any issues running Sykube on your machine or if you have any more questions feel free to ask.

**Comment by vsoch** [comment]
Thanks @Nienkeeijsvogel ! I have this in my TODO list - it's a bit downs so I can't promise to get to it immediately, but with SC next week (and me not going) and then the holidays I will definitely be able to make time. Thank you for updating the docs!

**Comment by jazzlyj** [comment]
@Nienkeeijsvogel  and @vsoch  I am a little late to this thread but very interested.
I am just now installing singularity 3.2.0 on ubuntu 22.04.1
will attempt to install branch v1.0.0-beta.8 of https://github.com/Nienkeeijsvogel/singularity-cri
and then skykube

doh! just installed singularity 3.2.0
gonna scratch that and go with the latest 3.10.4 as of writing this.

also I am not a "developer".
I am on the SysAdmin side of things and very interested in making use of singularity containers (for their security vs docker) orchestrated by K8s for HPC workloads.

**Comment by Nienkeeijsvogel** [comment]
@jazzlyj Sykube works without installing the singularity-cri because its already in the image!
Currently for Sykube to run, Ubuntu 22.04 requires booting in cgroupv1  with grub.
I explained how in the README of Sykube.
Let me know if you run into any issues:)

Sykube works with Singularity v3.5.2 and Kubernetes v.1.20.9 in its container image, so if you want to use the Singularity-cri (branch v1.0.0-beta.8) without Sykube, installing these versions on the host will surely work.

**Comment by jazzlyj** [comment]
@Nienkeeijsvogel 
I installed it last night and it didnt work.
I forgot to do the step you said about the changing grub, will try today this am.

I had installed singularity 3.10.4 and then the cri.
then installed sykube and saw it installs sing 3.5.2 and k8s 1.20.9 
oh well. I dont think I fully understood your docs or probably misread them. it was late :)

anyhow I really appreciate the work you have done on this!!!
will give it another try this am and keep you posted.

again many thanks!!!

Also I am happy to contribute however I can, but as I said I dont know go code.
I am great at writing documentation though if that can help.

**Comment by jazzlyj** [comment]
@Nienkeeijsvogel 
I am up and running after fixing grub.

Awesome!!!!

Thank you so much!!

**Comment by Nienkeeijsvogel** [comment]
@jazzlyj 
Great!

i updated the README of Sykube as well as Singularity-cri so its also clear how to run the cri without Sykube.
